Add filters to customize additional and address checkout fields in frontend

alexmigf opened the pull request:
### Changes proposed in this Pull Request:

This Pull Request introduces new filters to allow developers to customize the rendering of additional and address checkout fields in frontend. Specifically:

- Added `woocommerce_checkout_render_additional_field` filter in the `render_additional_field()` method to allow customization or removal of additional fields registered using `woocommerce_register_additional_checkout_field()`.
- Added `woocommerce_checkout_render_address_field` filter in the `render_address_fields()` method to provide the same level of control over address-related fields.

Closes #51949

### How to test the changes in this Pull Request:

1. Register additional fields using `woocommerce_register_additional_checkout_field()`.
2. Implement the `woocommerce_checkout_render_additional_field` filter to modify or remove these fields.
3. Verify that the filter successfully customizes the rendering of each additional field in the checkout.
4. Implement the `woocommerce_checkout_render_address_field` filter to modify or remove address fields.
5. Verify that the filter successfully customizes the rendering of address fields on the account page.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch

#### Type

-   [ ] Add - Adds functionality

#### Message 

Add filters `woocommerce_checkout_render_additional_field` and `woocommerce_checkout_render_address_field` to enable customization of additional and address checkout fields in frontend.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

This Pull Request adds new filters for customizing the rendering of additional and address checkout fields in the frontend.

</details>

alexmigf committed:
Adds new filters to render additional fields

#51949

alexmigf committed:
Update CheckoutFieldsFrontend.php

alexmigf committed:
Update CheckoutFieldsFrontend.php

alexmigf committed:
Merge branch 'trunk' into trunk

wavvves reviewed (changes requested):
Hi @alexmigf, thank you for the contribution!
Unfortunately the PR will need a few things before being reviewed:
1. Add filter documentation ([WP hook guidelines](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#4-hooks-actions-and-filters) and [Woo Developer Docs](https://developer.woocommerce.com/docs/how-to-add-actions-and-filters/) for guidance.
2. Add more detailed testing steps, eg with code snippets examples.
3. Provide a changelog file. This is not required for review, but it will be required for merging once approved.
4. There seems to be some [linting issues](https://github.com/woocommerce/woocommerce/actions/runs/11255944599/job/31921252458?pr=51951) being picked up by CI that need addressing.

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

alexmigf committed:
Update CheckoutFieldsFrontend.php

alexmigf committed:
Lint fix + changelog

alexmigf commented:
@wavvves 

Thank you for your feedback!

I've addressed the issues you raised.

Regarding the steps or code examples, as I [mentioned](https://github.com/woocommerce/woocommerce/issues/51949#issuecomment-2402651463) in the issue, our use case requires us to remove specific additional fields that we add to the address. Here's an example of how we handle it:

```php
add_filter( 'woocommerce_checkout_render_additional_field', function( $output, $field ) {
	return ( false !== strpos( $field['id'], 'wcnlpc/' ) ) ? '' : $output;
}, 10, 2 );

add_filter( 'woocommerce_checkout_render_address_field', function( $output, $field, $value, $key, $address_type, $_class ) {
	return ( false !== strpos( $field['id'], 'wcnlpc/' ) ) ? '' : $output;
}, 10, 6 );
```

This approach works well for us, effectively removing the fields from the Thank You page and in My Account addresses.

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

wavvves reviewed (changes requested):
>Add more detailed testing steps, eg with code snippets examples.
>There seems to be some [linting issues](https://github.com/woocommerce/woocommerce/actions/runs/11255944599/job/31921252458?pr=51951) being picked up by CI that need addressing.

There seems to be still [some linting issues](https://github.com/woocommerce/woocommerce/actions/runs/12235075238/job/35108060500?pr=51951), and would appreciate that testing steps could be more concrete with code examples, and if possible using a user friendly way to implement those (such as the Code Snippets extension).

In the meantime I'll be evaluating the filters and any alternate ways to achieve the result in the use case you mention. I see the field was left blank in the issue, have you considered any alternative way to achieve this without introducing new filters?

alexmigf commented:
> There seems to be still [some linting issues](https://github.com/woocommerce/woocommerce/actions/runs/12235075238/job/35108060500?pr=51951), and would appreciate that testing steps could be more concrete with code examples, and if possible using a user friendly way to implement those (such as the Code Snippets extension).

I’ll fix the linting issues. The example I shared [here](https://github.com/woocommerce/woocommerce/pull/51951#issuecomment-2431786287) should be enough for implementation. I can’t imagine other scenarios needing this filter, but if you want to manipulate that data, there probably isn’t another way to do it.

> In the meantime I'll be evaluating the filters and any alternate ways to achieve the result in the use case you mention. I see the field was left blank in the issue, have you considered any alternative way to achieve this without introducing new filters?

I couldn’t find alternatives then—only DOM manipulation with JS or CSS.

wavvves commented:
>The example I shared https://github.com/woocommerce/woocommerce/pull/51951#issuecomment-2431786287 should be enough for implementation.

Sorry, not sure I'm following this example. On the example you gave it is just to prevent them from showing on my account and thank you page right? Do you have a use case for customization?

I don't think you need new filters just for preventing them from displaying there.

For my account, in the edit addresses view (eg `/my-account/edit-address/shipping/`) you can tap into the `woocommerce_form_field_args`. Just be sure to add the filter with the `woocommerce_before_edit_address_form_[billing|shipping]` action and remove it with `woocommerce_after_edit_address_form_[billing|shipping]`, or else this will execute it with every `woocommerce_form_field()` call, potentially removing this from unwanted places.
Example:

```php
add_filter(
	'woocommerce_form_field_args',
	function( $field ) {
		if ( false !== strpos( $field['id'], 'wcnlpc/' ) {
			return null;
		}
		return $field;
	}
);
```

To remove it from `/my-account/edit-addresses/` I believe you can remove the action `woocommerce_my_account_after_my_address` attaching `CheckoutFieldsFrontend::render_address_fields` and re-add the action with your implementation. But since it's not for edition, why not just hide it with JS/CSS? Same goes for the Thank you page.

wavvves commented:
cc/ @opr do you have any thoughts on this?

opr reviewed (commented):
@wavvves the snippet you provided for `woocommerce_form_field_args` has a syntax error, it should be:

```php
add_filter(
	'woocommerce_form_field_args',
	function( $field ) {
		if ( false !== strpos( $field['id'], 'wcnlpc/' ) ) {
			return null;
		}
		return $field;
	}
);
```

As for the approach here, I want to take some time to consider the filter approach vs. a config at field registration time.

We want to be able to control rendering in:

1. Order confirmation screen
2. Order details screen (`my-account/view-order/[id]`)
3. Account Details screen (`my-account/edit-account/`)
4. Address edit screen in view mode (`my-account/edit-address/`) 
5. Address edit screen in edit mode (`my-account/edit-address/billing`) 

We can apply the `woocommerce_checkout_render_additional_field` filter to solve 1, 2, and 4, a developer would need to use that _and_ another filter to hide the fields from 3 and 5.

If we added an option at field registration time, we could control visibility internally and the developer would only have to worry about one setting in one place. This would be ideal. The problem with this approach is that @senadir is working on a visibility parser for field visibility in the Checkout _form_ so without knowing how this looks, we run the risk of introducing confusing or conflicting APIs at field registration time.

If we do want to continue with the filters, I recommend we drop the `render_additional_field` function from `AbstractOrderConfirmationBlock` and use the one in `CheckoutFieldsFrontend` instead (we can get this with: `Automattic\WooCommerce\Blocks\Package::container()->get( CheckoutFieldsFrontend::class );` and the `render_additional_field` method would have to be made public), I would also recommend dropping `woocommerce_checkout_render_address_field` and just use `woocommerce_checkout_render_additional_field` again, the API is the same so I don't see the value of having two filters.

I would also use @wavvves suggestion of using `woocommerce_form_field_args` to control field visibility in the edit view.

wavvves commented:
@alexmigf seems like the suggested alternatives cover the usage of the introduced filters. I'll close this PR for now, please open a new Issue if anything is missed to fulfill your use-case.

opr commented:
@wavvves I don't think this should be closed right away, it is still a valid approach, I just think we should discuss more about the use of filters vs. setting visibility at field registration. I would like @senadir to input on which approach is preferred.

My opinion is that these filters are actually preferred as doing it at field registration time could make a poor UX when it comes to differentiating this and the conditional/visibility system Nadir is working on.

wavvves commented:
Thanks for the heads up @opr, I misunderstood your comment. Let's then take in consideration what the two filters would enable vs what can be achieved through other methods that fit the use case described here. 
Changing content is one, but imo at html render time seems to hinder the full potential. Earlier, while the data is structured and easily modifiable might pose a stronger case.
I reopened the PR, I also think it's valuable to hear @senadir 👍

alexmigf committed:
Merge branch 'trunk' into trunk

alexmigf commented:
We're already handling the edit address functionality in the **My Account** section using `woocommerce_address_to_edit`, so that part isn't an issue for now. My concern is with the **address display** on the **My Account page**, the **Order View page**, and the **Thank You page**. While adding JS/CSS could work as a quick fix, it feels more like a workaround rather than a proper solution.

alexmigf committed:
Merge branch 'trunk' into trunk

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

opr commented:
@alexmigf hi - I'm wondering if this PR is still needed given we have conditional visibility options available:

[Conditional visibility and validation via JSON Schema](https://developer.woocommerce.com/docs/block-development/extensible-blocks/cart-and-checkout-blocks/additional-checkout-fields/#conditional-visibility-and-validation-via-json-schema)

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk

alexmigf committed:
Merge branch 'woocommerce:trunk' into trunk
